language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,362 @@
+}
+
+// A complete or partial instantiation of a function
+pub struct FunctionSpec {
+    pub handle: FunctionHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+// A complete or partial instantiation of a field (or the type of it).
+// A FieldSpec points to a generic FieldHandle and the owner type instantiation.
+pub struct FieldSpec {
+    pub field_spec: FieldHandleIndex,
+    pub instantiation: SignatureIndex,
+}
+
+
+//
+// Definitions
+//
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructDefinition {
+    pub struct_handle: StructHandleIndex,
+    pub layout: StructLayout,
+}
+
+// Fields are inlined and each struct carries the fields it owns.
+// The instructions that use fields ([imm]borrow_field) now have to specify
+// the type that owns the field and the field index.
+// For example, if you are referring to a field of `Cup<T>` instantiated
+// let c: cup<int> = ...;
+// let tea_type = &c.f // f typed T
+// here we want to put in the instruction borrow_field (Cup<int>, 1) // assuming f is field 1
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum StructLayout {
+    Native,
+    Declared(Vec<FieldDefinition>),
+}
+
+// no reason to carry the owner in the FieldDefinition now, as it is implicit
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldDefinition {
+    pub name: IdentifierIndex,
+    pub signature: Type,
+}
+
+// we want a better way to define native functions. That encoding is in `Code` which can
+// be native or move bytecoces
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct FunctionDefinition {
+    pub function: FunctionHandleIndex,
+    pub flags: u8, // this is just the `public` flag now
+    pub acquires_global_resources: Vec<StructDefinitionIndex>,
+    pub code: Code,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Code {
+    Native,
+    Declared(CodeUnit),
+}
+
+pub struct CodeUnit {
+    // ... max_stack_size is gone...
+
+    // locals carry only the locals signature without the arguments.
+    // However there is a single instruction that access locals uniformly,
+    // that is, as if there was a single vector of arguments + locals.
+    // The compiler (and generally clients) are supposed to do the math properly
+    pub locals: SignatureIndex,
+    pub code: Vec<Bytecode>,
+}
+
+//
+// Bytecodes
+//
+
+// The main issue to digest here is that every bytecode that deals with generic is now
+// split in 2 versions. The non generic and the generic. That has some advantages
+// we have talked about and we may touch on more
+pub enum Bytecode {
+    Pop,
+    Ret,
+    Nop, // should we have one?
+
+    // the branch instructions pending a structured control flow model we may want to provide
+    BrTrue(CodeOffset),
+    BrFalse(CodeOffset),
+    Branch(CodeOffset),
+
+    // load value on stack
+    LdU8(u8),
+    LdU64(u64),
+    LdU128(u128),
+    LdU8Vector(U8VectorPoolIndex),
+    LdAddr(AddressPoolIndex),
+    LdTrue,
+    LdFalse,
+
+    // random operators
```

Comment:
Hopefully we will update these comments :p

Reply:
LOL, yeah no one of those comments will be in the final version, or better, only those that make sense.
To be concrete the idea is to move the changes piece by piece in `file_format.rs` so the final "version" will be more in line with that file than what it is here.
The arrangement here is more for review than anything else